### ISSUE TITLE
PageLayout validation

### ISSUE THREAD

**Issue by chuckharmston** [Threespot/django-pagemanager#1]
It should be possible to validate the usage of PageLayouts.

Examples of validations that should be allowed to be performed:
- Layout of type A can only be a child of PageLayout of type B
- There can only be two pages that use PageLayout of type C.

I would investigate the possibility of doing this via a special class method of PageLayout.

**Comment by mazelife** [comment]
I like this idea. Can we come up with a complete list of the types of validation rules? So far:
- Max number of layouts allowed
- Required parent

**Comment by chuckharmston** [comment]
Rather than define specific categories of rules, I propose that we add an overridable classmethod to the PageLayout class in which any sort of rule can be defined. Well, that's not exactly true, there are two qualitatively different sorts of validation rules that we need to account for:
1) Rules to validate a PageLayout when it has been created but not saved (i.e. while processing the PageAdmin add_view POST). This will include the sorts of rules outlined above.
2) Rules to validate a PageLayout's data.

The latter case can be handled by PageLayoutAdmin.full_clean (and it's constituent methods). The former case is trickier, as it must live independently of any row-level validation. Here's where the class method comes in:

``` python
class TestLayout(PageLayout):
    """
    Demonstrating an example of PageLayout validation
    """

    @classmethod
    def validate_layout(cls, page):

        # Verify that TestLayout is always a child of ParentTestLayout
        parent_content_type = ContentType.objects.get_for_model(ParentTestLayout)
        if page.get_parent().layout_type != parent_content_type:
            raise ValidationError("Test layouts must always be children of Parent test layouts"

        # Verify that no more than 2 TestLayout instances exist
        count = cls.objects.all()
        if count > 2:
            raise ValidationError("There may not be more than two test layouts")

```

Thoughts? It will take some finessing to get implementation right, but it seems like a fairly clean API. We can also add shortcuts for the common validation types, such as the ones we've outlined above.

**Comment by chuckharmston** [comment]
Addressed in [master/4c7a687](https://github.com/Threespot/django-pagemanager/commit/4c7a6873ccb28bd306e920eb9e0a3c13853c0114).
